cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -350,34 +348,15 @@ func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc
 //  users don't have to restart their controller manager if they change the apiserver.
 // Until we get there, the structure here needs to be exposed for the construction of a proper ControllerContext.
 func GetAvailableResources(clientBuilder controller.ControllerClientBuilder) (map[schema.GroupVersionResource]bool, error) {
-	var discoveryClient discovery.DiscoveryInterface
-
-	var healthzContent string
 	// If apiserver is not running we should wait for some time and fail only then. This is particularly
 	// important when we start apiserver and controller manager at the same time.
-	err := wait.PollImmediate(time.Second, 10*time.Second, func() (bool, error) {
-		client, err := clientBuilder.Client("controller-discovery")
-		if err != nil {
-			glog.Errorf("Failed to get api versions from server: %v", err)
-			return false, nil
-		}
-
-		healthStatus := 0
-		resp := client.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
-		if healthStatus != http.StatusOK {
-			glog.Errorf("Server isn't healthy yet.  Waiting a little while.")
-			return false, nil
-		}
-		content, _ := resp.Raw()
-		healthzContent = string(content)
-
-		discoveryClient = client.Discovery()
-		return true, nil
-	})
+	client := clientBuilder.ClientOrDie("controller-discovery")
+	err := genericcontrollerconfig.WaitForAPIServer(client, 10*time.Second)
```

Comment:
Promote this out into the `CreateContext` method where it is called.  Make it more obvious that we wont'/can't produce a context on a bad API server.

This bit of code about available resources is likely to be a target for refactoring.

Reply:
good idea